cpp/locality/LinkCell.cc

```diff
@@ -387,32 +418,49 @@ NeighborPoint LinkCellQueryBallIterator::next()
                 }
             }
 
-            // Determine the next neighbor cell to consider. We're done if we
-            // reach a new shell and the closest point of approach to the new
-            // shell is greater than our rcut.
-            ++m_neigh_cell_iter;
+            bool out_of_range = false;
 
-            if ((m_neigh_cell_iter.getRange() - m_extra_search_width) * m_linkcell->getCellWidth() > m_r)
+            while (true)
             {
-                break;
+                // Determine the next neighbor cell to consider. We're done if we
+                // reach a new shell and the closest point of approach to the new
+                // shell is greater than our rcut.
+                ++m_neigh_cell_iter;
+
+                if ((m_neigh_cell_iter.getRange() - m_extra_search_width) * m_linkcell->getCellWidth() > m_r)
+                {
+                    out_of_range = true;
+                    break;
+                }
+
+                // In cases where we need to check the entire box, make sure
+                // that we don't check the same cell on the positive and
+                // negative sides of the IteratorCellShell cube.
+                if (check_double_range_shell(m_linkcell, *m_neigh_cell_iter))
+                    continue;
```

Comment:
I think this check is actually not necessary, since we're tracking (and skipping) the previously-searched cells in `m_searched_cells`. If you agree, @vyasr, I'd consider removing it. (The tests pass if it's removed.)

Reply:
I agree, you can remove this. I think I understand what your point (1) was referring to. I probably just overlooked this possibility for ball queries; it would have come up naturally in a sparse system for k-nearest neighbor queries since cutoffs don't matter there, but I didn't test large enough cutoffs to see that scenario with ball queries. However, it's also clearly not strict enough to catch e.g. the test case I posted in this PR's thread because you have to search that cell, you just need to avoid searching it twice.

Reply:
Removed.